internal/venice-common/src/main/java/com/linkedin/venice/utils/AvroRecordUtils.java

```diff
@@ -0,0 +1,235 @@
+            }
+            break;
+          case ENUM:
+          case FIXED:
+          case BYTES:
+          case RECORD:
+            throw new VeniceException(
+                "Default value for field: " + field.name() + " with type: " + fieldType + " is not supported");
+        }
+      }
+    }
+
+    return recordType;
+  }
+
+  /**
+   * Checks if it is possible for some value to be modified to adapt to the provided schema type.
+   * @param expectedSchemaType The type {@link Schema.Type} that the value needs to be adapted to.
+   * @return {@code true} if a value can be modified to adapt to the provided schema type; {@code false} otherwise.
+   */
+  private static boolean canSchemaTypeBeAdapted(Schema.Type expectedSchemaType) {
+    switch (expectedSchemaType) {
+      case ENUM:
+      case FIXED:
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case NULL:
+        return false;
+      case RECORD:
+      case ARRAY:
+      case MAP:
+      case UNION:
+        return true;
+      default:
+        // Defensive coding
+        throw new VeniceException("Unhandled Avro type. This is unexpected.");
+    }
+  }
+
+  /**
+   * Tries to adapt the {@param datum} to the {@param expectedSchema}.
+   * The following steps are followed:
+   * 1. If the schema type doesn't allow for adaptation, return the same value that was passed in input.
+   * 2. If the schema type allows for adaptation, then
+   *     2a. If the value doesn't specify a value for any field, the default value is used
+   *     2b. If a field is mandatory, but no default values are specified, then an Exception is thrown
+   *
+   * @param expectedSchema The type {@link Schema} that the value needs to be adapted to.
+   * @param datum The value that needs to be adapted to the specified schema
+   * @return
+   * @throws VeniceException
+   */
+  public static Object adaptToSchema(Schema expectedSchema, Object datum) throws VeniceException {
+    Schema.Type expectedSchemaType = expectedSchema.getType();
+    switch (expectedSchemaType) {
+      case ENUM:
+      case FIXED:
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case NULL:
+        return datum;
+      case RECORD:
+        return adaptRecordToSchema(expectedSchema, datum);
+      case ARRAY:
+        return adaptListToSchema(expectedSchema, datum);
+      case MAP:
+        return adaptMapToSchema(expectedSchema, datum);
+      case UNION:
+        return adaptUnionToSchema(expectedSchema, datum);
+      default:
+        // Defensive coding
+        throw new VeniceException("Unhandled Avro type. This is unexpected.");
+    }
+  }
+
+  private static Object adaptRecordToSchema(Schema expectedSchema, Object datum) {
+    if (!(datum instanceof IndexedRecord)) {
```

Comment:
Hmm what does the IndexedRecord requirement mean? Is all Record in Avro an IndexedRecord?

Reply:
Yes, both `GenericRecord` and `SpecificRecord` extend `IndexedRecord`
```
public interface GenericRecord extends IndexedRecord

public interface SpecificRecord extends IndexedRecord
```